src/dex/uniswap-v3/uniswap-v3.ts

```diff
@@ -0,0 +1,824 @@
+  UNISWAPV3_SUBGRAPH_URL,
+  UNISWAPV3_TICK_GAS_COST,
+} from './constants';
+import { DeepReadonly } from 'ts-essentials';
+import { uniswapV3Math } from './contract-math/uniswap-v3-math';
+import { Contract } from 'web3-eth-contract';
+import { AbiItem } from 'web3-utils';
+
+type PoolPairsInfo = {
+  token0: Address;
+  token1: Address;
+  fee: string;
+};
+
+const UNISWAPV3_QUOTE_GASLIMIT = 200_000;
+
+export class UniswapV3
+  extends SimpleExchange
+  implements IDex<UniswapV3Data, UniswapV3Param>
+{
+  readonly eventPools: Record<string, UniswapV3EventPool | null> = {};
+
+  readonly hasConstantPriceLargeAmounts = false;
+  readonly needWrapNative = true;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(UniswapV3Config);
+
+  logger: Logger;
+
+  private uniswapMulti: Contract;
+
+  constructor(
+    protected network: Network,
+    dexKey: string,
+    protected dexHelper: IDexHelper,
+    protected adapters = Adapters[network] || {},
+    readonly routerIface = new Interface(UniswapV3RouterABI),
+    readonly quoterIface = new Interface(UniswapV3QuoterABI),
+    protected config = UniswapV3Config[dexKey][network],
+    protected poolsToPreload = PoolsToPreload[dexKey][network] || [],
+  ) {
+    super(dexHelper, dexKey);
+    this.logger = dexHelper.getLogger(dexKey);
+    this.uniswapMulti = new this.dexHelper.web3Provider.eth.Contract(
+      UniswapV3MultiABI as AbiItem[],
+      this.config.uniswapMulticall,
+    );
+    // To receive revert reasons
+    this.dexHelper.web3Provider.eth.handleRevert = false;
+
+    // Normalise once all config addresses and use across all scenarios
+    this.config = this._toLowerForAllConfigAddresses();
+  }
+
+  get supportedFees() {
+    return this.config.supportedFees;
+  }
+
+  getAdapters(side: SwapSide): { name: string; index: number }[] | null {
+    return this.adapters[side] ? this.adapters[side] : null;
+  }
+
+  getPoolIdentifier(srcAddress: Address, destAddress: Address, fee: bigint) {
+    const tokenAddresses = this._sortTokens(srcAddress, destAddress).join('_');
+    return `${this.dexKey}_${tokenAddresses}_${fee}`;
+  }
+
+  async initializePricing(blockNumber: number) {
+    // This is only for testing, because cold pool fetching is goes out of
+    // FETCH_POOL_INDENTIFIER_TIMEOUT range
+    await Promise.all(
+      this.poolsToPreload.map(async pool =>
+        Promise.all(
+          this.config.supportedFees.map(async fee =>
+            this.getPool(pool.token0, pool.token1, fee, blockNumber),
+          ),
+        ),
+      ),
+    );
+  }
+
+  async getPool(
+    srcAddress: Address,
+    destAddress: Address,
+    fee: bigint,
+    blockNumber: number,
+  ): Promise<UniswapV3EventPool | null> {
+    let pool =
+      this.eventPools[this.getPoolIdentifier(srcAddress, destAddress, fee)];
+    if (pool === undefined) {
+      const [token0, token1] = this._sortTokens(srcAddress, destAddress);
```

Comment:
I see in types that `pool` may be `UniswapV3EventPool | null`, but it is checked against `undefined`. Don't know if it is intentional, but looks strange

Reply:
Ah, remembered it. `undefined` - is for uninitialized pool. In that case, may you please, add comment about that? I think it is not intuitive